Adding paradrop function

joshvelson opened the pull request:
This is a function called from a waypoint of a group piloting an aircraft.  It inserts 1-2 waypoints directly ahead of the waypoint that calls it. Once the new waypoint is reached, the aircraft conducts a rapid paradrop of groups of infantry.
  
The main arguments are the group piloting the vehicle, an array of group(s) to be dropped, and a marker denoting the drop target, with optional arguments for specifying drop characteristics or subsequent aircraft behavior.  Several default drop options are provided, corresponding to combat drops, training drops, HALO drops and civilian skydiving drops.  Can be called with the group containing the pilot as the group to be dropped also, though obviously the vehicle will be uncrewed afterwards.

Script uses getOut rather than the "Eject" action because the latter option requires the unit to go through the ejection animation.  This causes the paradrop to take about 4 seconds per unit and makes paradropping impractical for dropping significant forces.

getOut pushes the unit out instantly, with no delay.  However, it doesn't spawn a parachute and when looped also creates a "blob" of infantry leaving the vehicle at the beginning of the run.  Consequently this function has an adjustment factor used to ensure that infantry are dropped in sequence rathe than all at once.  The frequency for a plane's (speed > 250 kph) is different from the one for helicopters.

Tested with helicopters (Mi-8T) and fixed wing transports (C-130, C-47) with individual and multiple groups.  Jump frequency is tuned to allow a C-47 going 330 kph to unload an entire troop load (48 passengers) in a 1km drop run.  Mission file available upon request. 

Credit given to jonpas and veteran29 who both helped immensely during development.

jonpas reviewed (changes requested):
Code style:
- `[150, 100, 140], [...` (spaces)
- `if (` (space)
- `params ["_pilotGroup", "_dropGroups", ...]` (spaces)

I don't see the need to work with groups at all. From vehicle you can do `group (driver _vehicle)` to get the group of the pilot to do waypoints on. And you can do `fullCrew [_vehicle, "cargo"]` to get all passengers - then you can get rid of _all_ the group logic with delays and have a direct `_jumpDelay = _forEachIndex * _jumpFreq`.

Additionally, your current solution with `_paraGroups select {local leader _x}` will break as soon as one group is of different locality (eg. one group on server, one on HC), as your indices for `_groupAdjust` will suddenly be wrong.

I played around with it a bit, and this seems to work equally or slightly better than your version when it comes to dropping itself. Here is the branch with my changes (pretty much same as this review, I still posted the review to have a diff for you to look at; review does not have the above no-groups version, my branch does): https://github.com/Theseus-Aegis/MissionTemplate/blob/87842bc15f2b5078d8d1aca905a677619daf129e/scripts/fn_paradrop.sqf

Waypoints still seem broken, the waypoint applied is way off the marker, tested on Altis with @TyroneMF. I don't know if this worked with just your example well, and that's fine if it works for your mission, but at this moment, it does _not_ work globally. Tyrone will post some images here.

TyroneMF commented:
![107410_20220217124014_1](https://user-images.githubusercontent.com/29175040/154494258-b0de1151-c476-4fff-b1eb-082f8c4373ea.png)
![107410_20220217124500_1](https://user-images.githubusercontent.com/29175040/154494265-b367338d-8fba-4a94-b958-e5b2a85bf411.png)

TyroneMF commented:
Drop zones are a little off.

TyroneMF commented:
![107410_20220217134539_1](https://user-images.githubusercontent.com/29175040/154494355-4228d04c-4926-4f1f-85d2-04104a5fa89c.png)

joshvelson commented:
Re: Groups, the biggest issue is that when working with a single aircraft, it is sometimes either (a) not feasible to drop all cargo in a single pass or (b) undesirable to drop all cargo in a single pass.  

For (a), suppose you have a C-130 that carries approximately 80 passengers.  Having a DZ long enough to drop all passengers in sequence is impractical, and it is better to do multiple passes calling each time with the groups to be dropped specified.  

For (b) suppose that you have a single aircraft and you desire to drop two different squads to two different DZs.  This again requires to you specify which group is dropped. 

I get that this makes it more complicated but I was trying to put together a function with all valid use cases.

> Additionally, your current solution with _paraGroups select {local leader _x} will break as soon as one group is of different locality (eg. one group on server, one on HC), as your indices for _groupAdjust will suddenly be wrong.

Valid, this was Phil's addition to deal with potential locality issues and I'm not sure how to deal with it otherwise. 

> Waypoints still seem broken

I'm not sure what's going on there; I'd have to see the function call in terms of parameters.  I definitely tested this on more than one map.  If I had to guess what was going on, it is likely that the parameters for the drop were set such that the DZ length (set as a custom number in `[_dropHeight,_dropLength,_chuteHeight]` was set very high.  I will do some independent testing of my own and revert.

joshvelson commented on scripts/fn_paradrop.sqf:
```diff
@@ -0,0 +1,198 @@
+private _vehicle = assignedVehicle (leader _pilotGroup);
+if(isNull _vehicle) exitWith {}; // then there's something seriously wrong here
+
+_vehicle flyInHeight _dropHeight; // set forced flying height
+if(_resetHeight) then {
+    [_vehicle] spawn {
+        params ["_vehicle"];
+        waitUntil {sleep 5; count (fullCrew [_vehicle,"cargo"]) == 0}; //checks every 5 seconds for all cargo exited
+        _vehicle flyInHeight 100; // this is the default value
+    };
```
Did not know to use CBA_fnc_waitUntilAndExecute, will do so from now on.

joshvelson commented on scripts/fn_paradrop.sqf:
```diff
@@ -0,0 +1,198 @@
+                private _chuteHeight = _this select 1;
+                unassignVehicle _man;
+                moveOut _man;
+                [_man, _chuteHeight] spawn {
+                    params ['_man','_chuteHeight'];
+                    waitUntil {sleep 0.5; ((getPosATL _man) select 2) <= _chuteHeight};
+                    _chute = createVehicle ['Steerable_Parachute_F', position _man, [], 0, 'FLY'];
+                    _chute setPos (getPosATL _man);
+                    _man moveInDriver _chute;
+                };
```
I assume grabbing and storing the ATL/ASL position is to prevent changes in position from causing issues during execution?  As before, I wasn't aware of `CBA_fnc_waitUntilAndExecute`, so I'll use that in the future.  

One issue that I see here @jonpas is that I understand why you'd put the `{_man distance _vehicle >= 15}` condition in to avoid collision with the vehicle, but it does mean I have to do some fairly extensive injury testing on the people being dropped.  This makes the minimum height for parachute deployment about 75m using this script, which is unrealistically low.  Could be that Arma just doesn't represent the deceleration very well in which case who cares, but otherwise it's a problem.

joshvelson replied:
Distance check changes it because `_man` exits `_vehicle` at the vehicle's current speed, meaning that opening up the distance depends at first mostly on Z-directional acceleration.  75m is the preset option 2 ("dangerous drop") height of 90m - 15m, so in fairness it is an exaggeration to say that it will make it 75m.  However, it does present a potential issue in terms of parachute deployment being much lower than we want it, strictly speaking.

One alternative way of resolving this I've seen is just to do something like `_man setVelocity [0,0,5]` right after the `getOut _man` execution which solves the distance issue as well, and makes drop zones a more accurate reflection of where troops land, but if it's ever used by players and not AI they'll have a... "jerky" experience and I chose not to do that.  

Ultimately I think what we have to do is retain your solution, adjust the height of the "dangerous drop" preset upwards and call it a day.  It's based on the Rhodesian C-47 drops which were typically around 90m and as low as 70m even though officially they were supposed to be around 120m; however, it won't be that much different at 120m.  Also, I don't expect it to be used much, as it's "dangerous" for a reason, the Rhodesians took 10% casualties from every drop just by virtue of landing.

jonpas replied:
Soldiers do not seem to retain velocity, or they lose it really really fast. I don't think it's an issue, but feel free to adapt however to make it work nicely.

joshvelson commented on scripts/fn_paradrop.sqf:
```diff
@@ -0,0 +1,198 @@
+        {
+            private _jumpDelay = _forEachIndex*_jumpFreq+(_groupAdjust select _groupCount)*_jumpFreq;
+            [{
+                private _man = _this select 0;
+                private _chuteHeight = _this select 1;
```
I don't understand the use of `params` in this context.  AFAICT the variables `_man`, `_chuteHeight` and `_vehicle` are not already defined inside of this eventHandler, unless `CBA_fnc_waitAndExecute` also transmits their original names instead of just an array of values.

jonpas replied:
> When used without argument, as shown in main syntax, internal variable `_this`, which is usually available inside functions and event handlers, is used as argument.

Meaning `params [...] == _this params [...]`.

Function arguments are always in `_this`, same for `CBA_fnc_waitAndExecute` arguments.

joshvelson replied:
OK - and to be sure I understand this, `CBA_fnc_waitUntilAndExecute`'s blocks are in the same scope as the `CBA_fnc_waitAndExecute` call?  Because what really confused me was using `params` _inside_ the script block you're passing to `CBA_fnc_waitUntilAndExecute`.

jonpas replied:
No, it's a different context. And each block in `CBA_fnc_waitUntilAndExecute` (condition and statement) is different.

`_this` is passed to `CBA_fnc_waitUntilAndExecute`, so it passes all arguments from current scope to it, where I use `params` again to split it out.

joshvelson replied:
Got it.

jonpas commented:
> Groups

Ok those are fair use-cases. The group handling can still be improved. Instead of dealing with group unit counts, just put together all units of the provided groups and continue working on it as units. That solves both the weird delay handling and the locality issue.

Something like this (weird merge of my version and your groups):
```sqf
private _paraGroups = _vehicle getVariable [QGVAR(paradropGroups), []];
// Waypoint statements execute globally, ignore any groups not local and exit if none local
_paraGroups = _paraGroups select {local (leader _x)};
if (_paraGroups isEqualTo []) exitWith {};

// Deal with units from here on
private _dropUnits = _paraGroups apply {units _x}
_dropUnits = flatten _dropUnits;

private _chuteHeight = _vehicle getVariable [QGVAR(paradropChuteHeight), []];
private _jumpFreq = [0.5, 0.7] select (speed _vehicle < 250);

{
    (units _x) allowGetIn false;

    private _jumpDelay = _forEachIndex * _jumpFreq;
    [{
        params ["_man"];

        unassignVehicle _man;
        moveOut _man;

        [{
            params ["_man", "_chuteHeight", "_vehicle"];
            (getPosATL _man) select 2 <= _chuteHeight && {_man distance _vehicle >= 15}
        }, {
            params ["_man", "_chuteHeight"];

            private _manPos = getPosATL _man;
            private _chute = createVehicle ["Steerable_Parachute_F", _manPos, [], 0, "FLY"];

            _man moveInDriver _chute;
        }, _this] call CBA_fnc_waitUntilAndExecute;
    }, [_x, _chuteHeight, _vehicle], _jumpDelay] call CBA_fnc_waitAndExecute;
} forEach _dropUnits;
```

joshvelson reviewed (commented):
Overall no issues with any of the revisions except for the alterations to `_fnc_dropParas` I noted.

joshvelson commented:
@jonpas so far as I am able to tell, your weird merge version preserves the unit organization in terms of keeping each unit dropped with its group.  If that's the case, then yes, your solution is more elegant and we should go with that.  Still confused on how you're using `params`, but we can discuss that on Slack if need be.

jonpas commented:
Yes, it will retain order, so group 1 will be dropped in its entirety before group 2 will start to be dropped, simply from how all units are being put into a single flat array.

joshvelson commented on scripts/fn_paradrop.sqf:
```diff
@@ -112,88 +118,97 @@ private _dropRunOrigin = getPosATL _vehicle; // calling position from leader wil
+private _dropZoneLocs = [_dropTarget, _dropRunOrigin, _dropLength] call _fnc_dzLocs;
+private _distanceToDZ = [_dropRunOrigin distance (_dropZoneLocs select 0), _dropRunOrigin distance (_dropZoneLocs select 1)];
+private _minDistance = selectMin _distanceToDZ;
+private _indexOfMin = _distanceToDZ find _minDistance;
```
If we had left out handling of `_minDistance < _dropLength` then the calculations of lines 142-144 could actually be replaced by a `switch` statement thus:
`_indexOfMin = switch (floor (_azimuth / 90)) do {
    case 0;
    case 1: {1;};
    case 2;
    case 3: {0;};
};`

Which would produce a mild optimization.

joshvelson replied:
Actually, forget this entirely.  I tried running this and it went extraordinarily poorly, and I finally diagnosed why.  The issue was never with my math, honestly, because I think I did it right the first time.  

The issue is that `getDir`'s azimuth finding behavior sucks.  The nominal syntax is `pos1 getDir pos2` where the azimuth is measured from `pos1` to `pos2`.  However, upon further testing, the function will unpredictably measure the azimuth from `pos2` to `pos1` instead regardless of function call.  For example, one of my groups `ghost_hawk_1` was approaching the DZ from the southeast quadrant  which should make its nominal azimuth to the target between 90 and 180 degrees.  Debug logs, however, show its measured azimuth at 309.927 - the RECIPROCAL course!   Hence the sine and cosine of the angle come out with the correct magnitudes but incorrect signs.  This effect appears to be completely unpredictable, thus we are left with the distance calculation intact.   

Here are some further test logs:

- xian_1: Approaches from NE quadrant, azimuth should be between 180 and 270, measured azimuth 237.775, valid.
- blackfish_1: Approach from NW quadrant, azimuth should be between 270 and 360, measured azimuth 137.985, implying approach from SE quadrant - reciprocal.
- chinook_1: Approach from SW quadrant, azimuth should be between 0 and 90, measured azimuth 53.4929, valid.

joshvelson replied:
So basically I'm closing this note.  Way to go, Bohemia.

joshvelson replied:
Upon further reflection, I've figured it out.  Bohemia's getDir uses compass direction and not trigonometric angle.  Compass direction has 0 as North and advances counterclockwise.  I can put together a switch statement as a parametric function to convert it and a separate one to get the correct minimum angle without calculations.

joshvelson commented:
@TyroneMF if you're willing, the current version has now fixed the waypointing issues.  It now incorporates some logic to determine which of the two points is closer and setting the closer of the two as the drop waypoint.  This ultimately stems from the fact that Bohemia's `getDir` function will unpredictably give the reverse azimuth that it should (i.e., from drop zone to vehicle rather than the reverse), so the magnitude of sine and cosine used to calculate the new WPs is unaffected but their signs are.

@jonpas I'm not sure what you want to do w/r/t the version you wrote, but I incorporated our discussed "weird merged" code and added a feature or two.  I would prefer this version make it into the repo if Rone's tests work out. 

My test results:

From Northern approach:
![image](https://user-images.githubusercontent.com/24276655/154609189-1da1d7e1-d96d-4b59-9f2a-1d8dec0816f6.png)
![image](https://user-images.githubusercontent.com/24276655/154609227-82a6fe77-982b-4c36-8c43-9b25a4443648.png)
![image](https://user-images.githubusercontent.com/24276655/154609137-0eedbf86-eeef-4afb-9770-8d6e4f7358fb.png)

From Southern Approach:
![image](https://user-images.githubusercontent.com/24276655/154609345-c4cc65e4-46f4-4aa9-a50a-744258a2a609.png)
![image](https://user-images.githubusercontent.com/24276655/154609393-35ac1c7e-e834-4990-99b1-52b8ed470413.png)

TyroneMF commented on scripts/fn_paradrop.sqf:
```diff
@@ -0,0 +1,219 @@
+ * This instead uses getOut, which also requires spawning parachutes.
+ * 
+ * Call from move waypoints of the group piloting with (isServer) check.
+ * Technically you could call the piloting group as the drop group,
+ * which would result in the vehicle becoming unoccupied after the pilot drops.
```
```suggestion
```
Add this part to known behaviour.

TyroneMF commented on scripts/fn_paradrop.sqf:
```diff
@@ -0,0 +1,219 @@
+ * WP from which it is called; if it is held up, then the drop WPs will not
+ * reflect the current azimuth.  If synchronizing WPs, recommend you sync WPs
+ * BEFORE the one that calls this function to avoid weird behavior, especially
+ * if the drop marker is close by.
+ *
```
```suggestion
 * if the drop marker is close by.
 * 5. If pilot group is added to drop groups the vehicle will become unoccupied.
 *
```

TyroneMF commented on scripts/fn_paradrop.sqf:
```diff
@@ -0,0 +1,219 @@
+ * 0: Pilot Group <GROUP>
+ * 1: Drop Groups <ARRAY>
+ * 2: Drop Target <MARKER>
+ * 3: Drop Mode <ARRAY> or <NUMBER> (Default 0)
+ * 4: Reset Height <BOOLEAN> (Default false) 
+ * 5: Disable Waypoint 2 (Default false)
```
```suggestion
 * 3: Drop Mode <ARRAY> or <NUMBER> (default: 0)
 * 4: Reset Height <BOOL> (default: false) 
 * 5: Disable Waypoint 2 <BOOL> (default: false)
```

TyroneMF commented on scripts/fn_paradrop.sqf:
```diff
@@ -0,0 +1,219 @@
+
+#define DROP_MODES [[150,1000,140],[105,1000,80],[250,1000,140],[4500,1000,900],[3000,1000,900]]
+
+params ["_pilotGroup","_dropGroups","_loc",["_dropMode",0],["_resetHeight",false],["_disableWP2",false]];
```
```suggestion
params ["_pilotGroup", "_dropGroups", "_loc", ["_dropMode", 0], ["_resetHeight", false], ["_disableWP2", false]];
```

TyroneMF commented on scripts/fn_paradrop.sqf:
```diff
@@ -0,0 +1,219 @@
+_dropData params ["_dropHeight", "_dropLength", "_chuteHeight"];
+
+private _vehicle = vehicle (leader _pilotGroup);
+if(isNull _vehicle) exitWith {}; // group not in vehicle
```
```suggestion
if (isNull _vehicle) exitWith {}; // group not in vehicle
```

TyroneMF commented on scripts/fn_paradrop.sqf:
```diff
@@ -0,0 +1,219 @@
+
+private _origHeight = (getPosATL _vehicle) select 2;
+_vehicle flyInHeight _dropHeight; // set forced flying height
+if(_resetHeight) then {
```
```suggestion
if (_resetHeight) then {
```

TyroneMF commented on scripts/fn_paradrop.sqf:
```diff
@@ -0,0 +1,219 @@
+
+// Obtain a random location within drop target
+// Between current position and drop target, find angle
+private _dropTarget = [_loc,true] call CBA_fnc_randPosArea;
```
```suggestion
private _dropTarget = [_loc, true] call CBA_fnc_randPosArea;
```

TyroneMF commented on scripts/fn_paradrop.sqf:
```diff
@@ -0,0 +1,219 @@
+
+    diag_log [_pilotGroup];
+private _fnc_dzLocs = {
+    params ["_dropTarget","_dropRunOrigin","_dropLength"];
```
```suggestion
    params ["_dropTarget", "_dropRunOrigin", "_dropLength"];
```

jonpas replied:
Not resolved.

TyroneMF commented on scripts/fn_paradrop.sqf:
```diff
@@ -0,0 +1,219 @@
+}; 
+
+private _dp = _pilotGroup addWaypoint [_dropZoneLocs select 1,-1,currentWaypoint _pilotGroup + 1,""];
+if(!_disableWP2) then {
```
```suggestion
if (!_disableWP2) then {
```

TyroneMF commented:
Review above, issue i saw is fixed.

jonpas commented on scripts/fn_paradrop.sqf:
```diff
@@ -0,0 +1,213 @@
+#include "..\script_component.hpp"
+/*
+ * Author: Josh Vee, Jonpas, Veteran29
+ * An all-purpose paradrop function
```
```suggestion
 * An all-purpose paradrop function.
```

jonpas commented on scripts/fn_paradrop.sqf:
```diff
@@ -0,0 +1,213 @@
+ * times for different groups loaded on the same vehicle, but optimally
+ * used either for single group on helicopter or multiple groups on planes.
+ *
+ * 
+ * 
```
```suggestion
```

jonpas commented on scripts/fn_paradrop.sqf:
```diff
@@ -0,0 +1,213 @@
+ * Optional Arguments: Drop Mode
+ * For convenience, this function has the following pre-set modes for drops:
+ * 0: Modern combat drop, 150m, parachutes deploy at 140m.
+ * 1: Dangerous combat drop, 90m, parachutes deploy at 80m. 
```
```suggestion
 * 1: Dangerous combat drop, 105m, parachutes deploy at 80m. 
```

jonpas commented on scripts/fn_paradrop.sqf:
```diff
@@ -0,0 +1,213 @@
+ * BEFORE the one that calls this function to avoid weird behavior, especially
+ * if the drop marker is close by.
+ *
+ * Optional Arguments: Drop Mode
```
Instead of using a number for this, it could also be a string, eg. "modern", "danger", "halo", "civ", etc. Just a different representation of it. Either is fine, both need good documentation.

joshvelson replied:
I understand; my intention was to use whatever convention we had in place so I defaulted to using the one from the artillery script where ammunition type is denoted by number.

jonpas commented on scripts/fn_paradrop.sqf:
```diff
@@ -0,0 +1,213 @@
+ * 2: Training drop, 250m, parachutes deploy at 140m.
+ * 3: HALO drop, 4.5km, parachutes deploy at 900m.
+ * 4: Civilian drop, 3km, parachutes deploy at 900m. 
+ * For custom drops, drop altitude, drop zone length, deployment altitude can be set.
```
```suggestion
 * For custom drops, drop altitude, drop zone length, deployment altitude can be set
 * by providing an array of those values, in meters.
```

jonpas commented on scripts/fn_paradrop.sqf:
```diff
@@ -0,0 +1,213 @@
+ */
+
+
+#define DROP_MODES [[150,1000,140],[105,1000,80],[250,1000,140],[4500,1000,900],[3000,1000,900]]
```
```suggestion
#define DROP_MODES [[150, 1000, 140], [105, 1000, 80], [250, 1000, 140] ,[4500, 1000, 900], [3000, 1000, 900]]
```

jonpas commented on scripts/fn_paradrop.sqf:
```diff
@@ -0,0 +1,213 @@
+private _dropData = [];
+// first process dropMode
+if (_dropMode isEqualType 0) then {
+    _dropData = DROP_MODES select _dropMode;
```
```suggestion
    if (_dropMode < 0 || {_dropMode > 4}) then {
        ERROR_1("Paradrop mode %1 invalid - defaulting to 0!",_dropMode);
        _dropMode = 0;
    };
    _dropData = DROP_MODES select _dropMode;
```

jonpas commented on scripts/fn_paradrop.sqf:
```diff
@@ -0,0 +1,213 @@
+_dropData params ["_dropHeight", "_dropLength", "_chuteHeight"];
+
+private _vehicle = vehicle (leader _pilotGroup);
+if(isNull _vehicle) exitWith {}; // group not in vehicle
```
```suggestion
if (isNull _vehicle) exitWith {ERROR("Paradrop pilot group not in vehicle!")};
```

joshvelson replied:
An interesting circumstance that occurred in one of my tests due to my own error was that you can call paradrops from other vehicles, e.g. plane A's pilots are told to drop plane B's groups and vice-versa.  This check won't fix it, but just something unexpected to note.

jonpas replied:
Easy fix:
```sqf
if (vehicle _pilotGroup != vehicle _dropGroup) exitWith {
    ERROR_1("Paradrop pilot group not in the same vehicle as drop group '%1'",_dropGroup)
};
```

This has nothing to do with that check on the suggested line though. :P

jonpas commented on scripts/fn_paradrop.sqf:
```diff
@@ -0,0 +1,213 @@
+// Between current position and drop target, find angle
+private _dropTarget = [_loc, true] call CBA_fnc_randPosArea;
+
+if (_dropTarget isEqualTo []) exitWith {};
```
```suggestion
if (_dropTarget isEqualTo []) exitWith {ERROR_1("Paradrop target for marker '%1' not found!",_loc)};
```

jonpas commented on scripts/fn_paradrop.sqf:
```diff
@@ -0,0 +1,213 @@
+
+if (_dropTarget isEqualTo []) exitWith {};
+
+    diag_log [_pilotGroup];
```
```suggestion
```

jonpas commented on scripts/fn_paradrop.sqf:
```diff
@@ -0,0 +1,213 @@
+
+params ["_pilotGroup", "_dropGroups", "_loc", ["_dropMode", 0], ["_resetHeight", false], ["_disableWP2", false]];
+
+if (_loc isEqualTo []) exitWith {};
```
```suggestion
if (_loc isEqualTo "") exitWith {ERROR("Paradrop given invalid marker!")};
```
Assuming this should check for an empty string - "no marker provided".

joshvelson replied:
I'm not sure it will be different as in my `diag_log` output an invalid marker was printed as `[any]` but as long as it works, it works.

jonpas replied:
That makes no sense, you expect a string, marker is just a string. At this point that string has no correlation to the marker, so you must have passed `nil` (undefined variable) to get `any`.

jonpas commented on scripts/fn_paradrop.sqf:
```diff
@@ -0,0 +1,213 @@
+    private _xDiff = (cos _theta) * _dropLength/2;
+    // set all coordinates based on DZ points in xy-plane; z coord does not matter
+    // double lengths on second DZ coordinate to ensure straight line flight
+    [[(_dropTarget select 0) + _xDiff,(_dropTarget select 1) + _yDiff,_dropHeight],
+        [(_dropTarget select 0) - _xDiff,(_dropTarget select 1) - _yDiff,_dropHeight]];
```
```suggestion
    _dropTarget params ["_dropTargetX", "_dropTargetY"];
    [
        [_dropTargetX + _xDiff, _dropTargetY + _yDiff, _dropHeight],
        [_dropTargetX - _xDiff, _dropTargetY - _yDiff, _dropHeight]
    ];
```

jonpas commented on scripts/fn_paradrop.sqf:
```diff
@@ -0,0 +1,213 @@
+    _dropZoneLocs = [_dropTarget,_dropRunOrigin, _dropLength] call _fnc_dzLocs;
+}; 
+
+private _dp = _pilotGroup addWaypoint [_dropZoneLocs select 1,-1,currentWaypoint _pilotGroup + 1,""];
```
```suggestion
private _dp = _pilotGroup addWaypoint [_dropZoneLocs select 1, -1,currentWaypoint _pilotGroup + 1, ""];
```

jonpas commented on scripts/fn_paradrop.sqf:
```diff
@@ -0,0 +1,213 @@
+
+private _dp = _pilotGroup addWaypoint [_dropZoneLocs select 1,-1,currentWaypoint _pilotGroup + 1,""];
+if (!_disableWP2) then {
+    private _de = _pilotGroup addWaypoint [_dropZoneLocs select 0,-1,currentWaypoint _pilotGroup + 2,""];
```
```suggestion
    private _de = _pilotGroup addWaypoint [_dropZoneLocs select 0, -1, currentWaypoint _pilotGroup + 2, ""];
```

jonpas commented on scripts/fn_paradrop.sqf:
```diff
@@ -0,0 +1,213 @@
+    } forEach _dropUnits;
+};
+
+_dp setWaypointStatements ["true",toString _fnc_dropParas];
```
```suggestion
_dp setWaypointStatements ["true", toString _fnc_dropParas];
```

jonpas commented on scripts/fn_paradrop.sqf:
```diff
@@ -0,0 +1,213 @@
+ * [pilot_group, [airborne_1], "drop_marker", 0, false, false] call TAC_Scripts_fnc_paradrop;
+ * [pilot_group, [airborne_1], "drop_marker", [200,1000,150]] call TAC_Scripts_fnc_paradrop;
+ */
+
```
```suggestion
```

jonpas commented on scripts/fn_paradrop.sqf:
```diff
@@ -0,0 +1,213 @@
+// Handling for when calling waypoint is within distance _dropLength
+if (_minDistance < _dropLength) then {
+    _dropLength = _minDistance - 100; //arbitrary distance to allow for a helicopter turn
+    if (_dropLength < 0) exitWith {ERROR_1("Waypoint too close to DZ for drop!")};
```
Actually, should this be an error? Or will this `exitWith` just exit the `if` statement above it and continue with the original `_dropZoneLocs`? If it shouldn't be an error, maybe a warning or entirely remove the message?

joshvelson replied:
You're right, even though `if` in the [docs](https://community.bistudio.com/wiki/exitWith) is not explicitly named as one of the local scopes that can be exited, it shows up in the examples.  Still, I think it should be an error, and the way to get it to properly be an error would actually be to code it as `exitWith {exitWith{ERROR_!("Waypoint too close to DZ for drop!")};};`

jonpas commented on scripts/fn_paradrop.sqf:
```diff
@@ -0,0 +1,213 @@
+    }, [_vehicle, _origHeight]] call CBA_fnc_waitUntilAndExecute;
+};
+
+// next process waypoints if waypoints desired. 
+// if waypoint is desired, then script will be run on setWaypointStatements
+// if not then script will be run in this function.
```
Is any of this still true? I am not sure what that comments means. :thinking:

joshvelson replied:
Yeah, these are obsolete.  I'll delete them.

jonpas commented on scripts/fn_paradrop.sqf:
```diff
@@ -0,0 +1,213 @@
+// if not then script will be run in this function.
+
+// Use addWayPoint for the group piloting the vehicle to add the drop waypoints
+private _dropRunOrigin = getPosATL _vehicle; // calling position from leader will call last position on the ground
```
```suggestion
private _dropRunOrigin = getPosATL _vehicle;
```
Comment not accurate, since we switched to `getPosATL` instead of `position` - it would report correctly but delayed. Doing it in the vehicle is always better.

joshvelson replied:
Noted, thanks.

jonpas commented on scripts/fn_paradrop.sqf:
```diff
@@ -0,0 +1,213 @@
+    private _de = _pilotGroup addWaypoint [_dropZoneLocs select 0,-1,currentWaypoint _pilotGroup + 2,""];
+};
+
+// Add the paradrop operation. To store variables, adds info to vehicle characteristics hash that can be referenced globally
```
```suggestion
// Add the paradrop operation. To store variables, adds info to vehicle name space that can be referenced later
```
It doesn't set those variables globally, just in the vehicle namespace (and only on the local client).

jonpas reviewed (commented):
Style, comments and documentation clarifications.

Last review from me, think I covered everything now.

jonpas commented on scripts/fn_paradrop.sqf:
```diff
@@ -0,0 +1,216 @@
+// first process dropMode
+if (_dropMode isEqualType 0) then {
+    if (_dropMode < 0 || {_dropMode > 4}) then {
+        ERROR("Paradrop mode %1 invalid - defaulting to 0!",_dropMode);
```
Why did you change this? I wrote it correctly, or it won't work at all (silent fail due to macros error handling).
```suggestion
        ERROR_1("Paradrop mode %1 invalid - defaulting to 0!",_dropMode);
```

jonpas commented on scripts/fn_paradrop.sqf:
```diff
@@ -0,0 +1,216 @@
+// Between current position and drop target, find angle
+private _dropTarget = [_loc, true] call CBA_fnc_randPosArea;
+
+if (_dropTarget isEqualTo []) exitWith {ERROR("Paradrop target for marker '%1' not found!",_loc)};
```
```suggestion
if (_dropTarget isEqualTo []) exitWith {ERROR_1("Paradrop target for marker '%1' not found!",_loc)};
```
Same here.

jonpas commented on scripts/fn_paradrop.sqf:
```diff
@@ -0,0 +1,216 @@
+params ["_pilotGroup", "_dropGroups", "_loc", ["_dropMode", 0], ["_resetHeight", false], ["_disableWP2", false]];
+
+if (_loc isEqualTo "") exitWith {ERROR("Paradrop given invalid marker!")};
+
```
Incorporating the fix for a bug you mentioned in previous review.
```suggestion

// make sure you delete these 2 lines later below in the script (and this comment too)! I just moved them here as early exit is better, and we have a use for it early here now
private _vehicle = vehicle (leader _pilotGroup);
if (isNull _vehicle) exitWith {ERROR("Paradrop pilot group not in vehicle!")};

private _origDropGroupsCount = count _dropGroups;
_dropGroups = _dropGroups select {_vehicle != vehicle (leader _x)};
if (count _dropGroups < _origDropGroupsCount) then {
    WARNING_1("Paradrop %1 drop group(s) skipped because they are not in the same vehicle as pilot group!",_origDropGroupsCount - count _dropGroup)
};
```

jonpas commented on scripts/fn_paradrop.sqf:
```diff
@@ -0,0 +1,216 @@
+    if (_dropLength > 0) then {
+        _dropZoneLocs = [_dropTarget,_dropRunOrigin, _dropLength] call _fnc_dzLocs;
+    } else {
+        diag_log("Warning: Vehicle too close to DZ!");
```
I am still not sure if being too close should _exit/abort_ the drop or just continue with the location found on L142?

If it should abort it, you should write it like so:
```sqf
private _withinDistance = _minDistance < _dropLength;
private _dropLength = _minDistance - 100; // arbitrary distance to allow for a helicopter turn
if (_withinDistance && {_dropLength < 0}) exitWith {
    ERROR_1("Paradrop vehicle '%1' too close to DZ!",_vehicle);
};
if (_withinDistance) then {
    _dropZoneLocs = [_dropTarget,_dropRunOrigin, _dropLength] call _fnc_dzLocs;
};
```
or like so:
```sqf
private _tooClose = false;
if (_minDistance < _dropLength) then {
    _dropLength = _minDistance - 100; //arbitrary distance to allow for a helicopter turn
    if (_dropLength < 0) exitWith {
        _tooClose = true;
    };
    _dropZoneLocs = [_dropTarget,_dropRunOrigin, _dropLength] call _fnc_dzLocs;
}; 

if (_tooClose) exitWith {ERROR_1("Paradrop vehicle '%1' too close to DZ!",_vehicle)};
```

If it should just continue with the earlier location, write this log like so and leave the rest of it as it is:
```sqf
WARNING_1("Paradrop vehicle '%1' too close to DZ!",_vehicle);
```

joshvelson replied:
Took final suggestion.

Lairdd1989 commented:
Any update on this being ready thought it was all good seeing its not would like to see a NCO use it first before running a contract using it.
@joshvelson

joshvelson commented:
@Lairdd1989 it's more or less ready, just me being snowed under by work right now.  I'll do as JJ suggests and remove the offending section, and in theory upon final review it should be good to go.

Mike-MF commented:
Closing, this no longer goes into the template but into `TAC Mods/Mission` instead.
